planet/cli/subscriptions.py

```diff
@@ -0,0 +1,126 @@
+    """Prints a sequence of JSON-encoded Subscription descriptions.
+
+    This implementation is only a placeholder. To begin, instead
+    of mocking calls to the Subscriptions API, we'll use a
+    collection of fake subscriptions (the all_subs fixture).
+    After we refactor we will change to mocking the API.
+
+    """
+    # Filter by status, like the Subscriptions API does.
+    if status:
+        select_subs = (sub for sub in _fake_subs if sub['status'] in status)
+    else:
+        select_subs = _fake_subs
+
+    filtered_subs = itertools.islice(select_subs, limit)
+    # End of placeholder implementation. In the future we will get
+    # the filtered subscriptions from a method of the to-be-written
+    # planet.clients.subscriptions.
+
+    # Print output to terminal, respecting the provided limit.
+    for sub in filtered_subs:
+        echo_json(sub, pretty)
+
+
+def parse_request(ctx, param, value: str) -> dict:
+    """Turn request JSON/file into a dict."""
+    if value.startswith('{'):
+        try:
+            obj = json.loads(value)
+        except json.decoder.JSONDecodeError:
+            raise click.BadParameter('Request does not contain valid json.',
+                                     ctx=ctx,
+                                     param=param)
+        if not obj:
+            raise click.BadParameter('Request is empty.', ctx=ctx, param=param)
+    else:
+        try:
+            with click.open_file(value) as f:
+                obj = json.load(f)
+        except json.decoder.JSONDecodeError:
+            raise click.BadParameter('Request does not contain valid json.',
+                                     ctx=ctx,
+                                     param=param)
+
+    return obj
+
+
+@subscriptions.command(name='create')
+@click.argument('request', callback=parse_request)
+@click.option('--pretty', is_flag=True, help='Pretty-print output.')
+@click.pass_context
+@translate_exceptions
+@coro
+async def create_subscription(ctx, request, pretty):
+    """Submits a subscription request and prints the API response.
+
+    This implementation is only a placeholder. To begin, instead
+    of mocking calls to the Subscriptions API, we'll use a
+    collection of fake subscriptions (the all_subs object).
+    After we refactor we will change to mocking the API.
+
+    """
+    # Begin fake subscriptions service. Note that the Subscriptions
+    # API will report missing keys differently, but the Python API
+    # *will* raise PlanetError like this.
+    missing_keys = {'name', 'delivery', 'source'} - request.keys()
+    if missing_keys:
+        raise PlanetError(f"Request lacks required members: {missing_keys!r}")
+
+    # Update the request with an id.
+    sub = dict(**request, id='42')
+    _fake_subs.append(sub)
+
+    # End fake subscriptions service. After we refactor we will get
+    # the "sub" from a method in planet.clients.subscriptions (which
+    # doesn't exist yet).
+
+    echo_json(sub, pretty)
```

Comment:
These functions are copied straight from the bodies of the tests in which they were born with almost no change.